src/poisson.f90

```diff
@@ -0,0 +1,509 @@
+!!                                                                                                    *
+!!    pre_conditioning_matrix(1:num_nodes) ......... Diagonal preconditioning matrix.                 *
+!!    rhs_vector(1:num_nodes) ...................... Right-hand side (RHS) vector.                    *
+!!    f_increment(1:num_nodes) ..................... Computed increment of f.                         *
+!!    boundary_index(1:num_nodes) .................. Index to set f_increment(ip) = 0 for Dirichlet   *
+!!                                                   boundary conditions.                             *
+!!    b(1:num_nodes) ............................... Stores beta as in Hestenes-Stiefel relation.     *
+!!                                                                                                    *
+!!----------------------------------------------------------------------------------------------------*
+module poisson
+      implicit none
+
+      integer, parameter :: mxp  = 10000
+      integer, parameter :: mxe  = 30000 
+      integer, parameter :: mxb  =  5000 
+      integer, parameter :: mxc  =   100
+
+      integer :: num_nodes, num_elements, num_boundary_points, num_sets, num_dirichlet_boundary_conditions, &
+                 num_neumann_boundary_conditions
+
+contains
+
+      subroutine open_file(file_name, status, file_io)
+            implicit none
+
+            integer, intent(out) :: file_io
+            character*120, intent(in) :: file_name
+            character*3, intent(in) :: status
+
+            integer :: iostat
+
+            print *, file_name
+            
+            open (file_io,                  &
+                  file=trim(file_name),     &
+                  status=status,            &
+                  IOSTAT=iostat)
+
+            if( iostat .ne. 0) then
+                  write(*,'(a)') ' *** Error when opening '//trim(file_name)
+            else
+                  write(*,'(/,a)') ' *** '//trim(file_name)//' opened'
+            end if
+      end subroutine open_file
+
+      !!-----------------------------------------------------------------------------*
+      !!                                                                             *
+      !!    "inp" reads the input data: triangular mesh and problem parameters.      *
+      !!                                                                             *
+      !!-----------------------------------------------------------------------------*
+      subroutine inp(element_to_node,vb2_index,coordinates,boundary_node_num,num_side_nodes,vb,vb1,vb2,file_io)
+            implicit none
+
+            integer, intent(out) :: element_to_node(3,mxp), vb2_index(mxe), coordinates(2, mxp), boundary_node_num(2,mxb), &
+                                    num_side_nodes(4,mxb)
+            real, intent(out)     :: vb(3,mxc), vb1(mxc), vb2(mxc)
```

Comment:
Arrays are statically allocated. We may want to change this to dynamic if we want to do experiments on performance.